### PR TITLE
Experimental talismans

### PR DESCRIPTION
**Pull request by svr333**
## Description
Implement 2 new talismans, Talisman of the Wise and Talisman of the Resurrected. Also refactors the Talisman class quite a bit.

Regarding my Testing results below, I dont know if Ender Talismans are supposed to work in inventories, otherwise the only thing needed to fix is the ability to right click the Ender variant of the Talisman of the Resurrected.

## Changes
- Everything from the old PR
- Add a return statement so it doesnt send 2 messages

## Testing Results
@ img

I did some thorough testing and came to the following results:
- Ender Talismans only work in enderchest, not in inventory
- Right clicking on the Ender Talisman of the Resurrected does not set a respawn point
- Both variants of the Talismans of the Anvil, Miner, Wise and Traveller work (I assume the rest works too but I didnt test all of them)

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by MasterPearl** [comment]
Hey. I am a little confused can someone explain how this plugin works? I am still learning.

**Comment by J3fftw1** [comment]
> Hey. I am a little confused can someone explain how this plugin works? I am still learning.

Join the discord this is not the place to ask

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/magical/talismans/ResurrectedTalisman.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        if (json != null) {
+            return parseLocationFromJsonObject(json);
+        } else {
+            return null;
+        }
+    }
+
+    @Nonnull
+    private JsonObject createJsonFromLocation(@Nonnull Location loc) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/magical/talismans/ResurrectedTalisman.java:80]: I would recommend just changing this. You can use BlockPosition and just store the long.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TalismanListener.java [hunk 2]
```diff
@@ -291,10 +303,47 @@ public void onBlockDropItems(BlockDropItemEvent e) {
...
+
+    @EventHandler
+    public void onPlayerDeath(PlayerDeathEvent e) {
+        Player player = e.getEntity();
+        DamageCause dmgCause = player.getLastDamageCause().getCause();
+        ItemStack talisman = Talisman.tryActivateAndGet(e, SlimefunItems.TALISMAN_RESURRECTED);
+
+        if (dmgCause == DamageCause.VOID && talisman != null) {
+            // This event is cancelled by the Talisman, so they don't actually die, we just teleport them after 2 ticks.
+            SlimefunPlugin.runSync(() -> PaperLib.teleportAsync(player, getSafeRespawnLocation(talisman, player)), 2);
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TalismanListener.java:325]: Any reason you delay this 2 ticks? This event will keep firing as the player is on half a heart and keeps getting damage from the void.

**Review by WalshyDev — CHANGES_REQUESTED** [review]
Quick look through. You can change the Location storage to just a simple `long` with BlockPosition. Makes it much easier and saves some nice memory and disk space.

**Comment by TheBusyBiscuit** [comment]
I think I will "cherry-pick" merge the Talisman of the Wise already, seems very useful and stable enough to make it into the /master/ branch.

**Comment by TheBusyBiscuit** [comment]
@svr333 For the Talisman of the Resurrected, you may need to make a new PR and revert the latest commit to `master`

**Pull request merged**